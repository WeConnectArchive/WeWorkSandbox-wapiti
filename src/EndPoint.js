import React, {Component} from 'react';
import cx from 'classnames';
import marked from 'marked';
import DangerousHTML from 'react-dangerous-html';
import {ApiRequest} from './ApiRequest';
import {ApiResponse} from './ApiResponse';

import style from './Endpoint.scss';
import stylef from './FontStyles.scss';

export class EndPoint extends Component {
  
  constructor(props) {
    super(props);
    this.state = { accept: '', expanded: false };
  }

  toggleExpand = (e) => {
    if (this.state.expanded){
      this.setState({
        expanded: false,
        accept: ''
      });
    }
    else {
      let accept ='';
      for(let respStatus in this.props.path.responses){
        for(let acceptContentType in (this.props.path.responses[respStatus]["content"])){
          accept = accept + acceptContentType + ', '
        }
      }
      accept = accept.replace(/,\s*$/, ""); // remove trailing comma
      this.setState({expanded: true, accept: accept});
    }
  }

  render() {
    return (
    <div  className={cx(style['m-endpoint'],stylef['regular-font'], style[this.props.path.method], style[(this.state.expanded?'expanded':'collapsed')])}>
      <div onClick={ this.toggleExpand } className={cx(style['head'],style[this.props.path.method],style[(this.state.expanded?'expanded':'collapsed')])}>
        <div className={cx(style['method'],style[this.props.path.method])} > {this.props.specType === 'rest' ? this.props.path.method : 'Method' } </div>
        <div className={cx(style['path'], style[this.props.path.deprecated?'deprecated':''])}> {this.props.path.path} </div>
        {this.props.path.deprecated && <span style={{ fontSize: '12px', textTransform: 'uppercase', fontWeight: 'bold', color: 'orangered', margin: '2px 0 0 5px' }}> deprecated </span>}
        <div className={cx(style["only-large-screen"])} style={{ minWidth: '60px', flex: '1' }}></div>
        <div className={cx(style["m-markdown-small"],style["descr"])}> {this.props.path.summary?<DangerousHTML html={marked(this.props.path.summary)} /> : ''}</div>
      </div>

      {this.state.expanded && (
      <div className={cx(style['body'],style[this.props.path.method])}>
        {(this.props.path.summary || this.props.path.description) && (
          <div className={cx(style["summary"])}>
            <div className={cx(style["m-markdown"], style["title"])}><DangerousHTML html={marked(this.props.path.summary)} /></div>
            {this.props.path.summary !== this.props.path.description && (
              <div className={cx(style["m-markdown"])}>
                {this.props.path.description?<DangerousHTML html={marked(this.props.path.description)} />:''}
              </div>
            )}
          </div>
        )}
        <div className={cx(style["req-resp-container"])}>
          <ApiRequest
            specType = {this.props.specType}
            method = {this.props.specType === 'rest' ? this.props.path.method : 'post'}
            path = {this.props.path.path}
            selectedServer = {this.props.selectedServer}
            parameters = {this.props.path.parameters}
            requestBody = {this.props.path.requestBody}
            accept ={this.state.accept}
          ></ApiRequest>
          <ApiResponse
            responses={this.props.path.responses}
            specType={this.props.specType}
          ></ApiResponse>
        </div>
      </div>
      )}

    </div>
    )
  }

}
