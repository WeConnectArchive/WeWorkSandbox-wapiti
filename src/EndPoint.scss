    .request{ border-width: 0 0 1px 0; }

    .only-large-screen {
      display:none;
    }

    .head .path{
      display: flex;
      font-family:var(--font-mono);
      font-size: var(--small-font-size);
      align-items: center;
      overflow-wrap: break-word;
      word-break: break-all;
    }

    .head .descr{
      font-size: var(--small-font-size);
      color:var(--light-fg);
      font-weight:400;
      align-items: center;
      overflow-wrap: break-word;
      word-break: break-all;
      display:none;
    }

    .m-endpoint.expanded{margin-bottom:16px; }
    .m-endpoint > .head{
      border-width:1px 1px 1px 5px;
      border-style:solid;
      border-color:transparent;
      border-top-color:var(--light-border-color);
      display:flex;
      padding:6px 16px;
      align-items: center;
      cursor: pointer;
    }
    .m-endpoint > .head.put:hover,
    .m-endpoint > .head.put.expanded{
      border-color:var(--put-color);
      background-color:var(--light-put-color);
    }
    .m-endpoint > .head.post:hover,
    .m-endpoint > .head.post.expanded{
      border-color:var(--post-color);
      background-color:var(--light-post-color);
    }
    .m-endpoint > .head.get:hover,
    .m-endpoint > .head.get.expanded{
      border-color:var(--get-color);
      background-color:var(--light-get-color);
    }
    .m-endpoint > .head.delete:hover,
    .m-endpoint > .head.delete.expanded{
      border-color:var(--delete-color);
      background-color:var(--light-delete-color);
    }
    .m-endpoint > .head.patch:hover,
    .m-endpoint > .head.patch.expanded{
      border-color:var(--patch-color);
      background-color:var(--light-patch-color);
    }
    .m-endpoint .body {
      flex-wrap:wrap;
      padding:16px 0px 0 0px;
      border-width:0px 1px 1px 5px;
      border-style:solid;
      box-shadow: 0px 4px 3px -3px rgba(0, 0, 0, 0.15);
    }
    .m-endpoint .body.delete{ border-color:var(--delete-color); }
    .m-endpoint .body.patch{ border-color:var(--patch-color); }
    .m-endpoint .body.put{ border-color:var(--put-color); }
    .m-endpoint .body.post{border-color:var(--post-color);}
    .m-endpoint .body.get{ border-color:var(--get-color); }

    .head .deprecated{
      text-decoration: line-through red;
    }

    .summary{
      padding:8px 8px;
    }
    .summary .title{
      font-size:calc(var(--title-font-size) + 2px);
      margin-bottom: 6px;
      word-break: break-all;
    }

    .method{
      padding:2px 5px;
      vertical-align: middle;
      height: 20px;
      line-height: 20px;
      min-width: 48px;
      border-radius: 2px;
      display:inline-block;
      font-size:var(--small-font-size);
      text-align: center;
      font-weight: bold;
      text-transform:uppercase;
      margin-right:5px;
    }
    .method.delete{ border: 2px solid var(--delete-color);}
    .method.patch{ border: 2px solid var(--patch-color); }
    .method.put{ border: 2px solid var(--put-color); }
    .method.post{ border: 2px solid var(--post-color); }
    .method.get{ border: 2px solid var(--get-color); }

    .req-resp-container{
      display: flex;
      margin-top:16px;
      align-items: stretch;
      flex-wrap: wrap;
      flex-direction: column;
      border-top:1px solid var(--light-border-color);
    }
    .request,
    .response{
      flex:1;
      min-height:100px;
      padding:16px 8px;
      overflow:hidden;
    }
    .request{
      border-width:0 0 1px 0;
      border-style:dashed;
    }
    .patch .request{
      border-color:var(--patch-color);
    }
    .put .request{
      border-color:var(--put-color);
    }
    .post .request{
      border-color:var(--post-color);
    }
    .get .request{
      border-color:var(--get-color);
    }
    .delete .request{
      border-color:var(--delete-color);
    }

   @media only screen and (min-width: 768px){
      .head .path{
        font-size: var(--regular-font-size);
        min-width:400px;
      }
      .head .descr{
        display: flex;
      }
      .head .m-markdown-small,
      .descr .m-markdown-small{
        display:block;
      }
      .only-large-screen{
        display:block;
      }
      .req-resp-container{
        flex-direction: var(--layout, row);
      }
      .request{
        border-width:0 1px 0 0;
        padding:16px 24px;
      }
      .response{
        padding:16px 24px;
      }
      .summary{
        padding:8px 24px;
      }
    }
