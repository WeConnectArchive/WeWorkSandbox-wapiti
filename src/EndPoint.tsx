import * as React from 'react';
import marked from 'marked';
import DangerousHTML from 'react-dangerous-html';
import {ApiRequest} from './ApiRequest';
import {ApiResponse} from './ApiResponse';
import './EndPoint.css'
  
interface EndPointProps {
    path: any;
    selectedServer: string;
    specType: string;
};

interface EndPointState {
    accept: string;
    expanded: boolean;
};

export class EndPoint extends React.Component<EndPointProps,EndPointState> {

  constructor(props: EndPointProps) {
    super(props);
    this.state = { accept: '', expanded: false };
  }

  toggleExpand = (e) => {
    if (this.state.expanded){
      this.setState({
        expanded: false,
        accept: ''
      });
    }
    else {
      let accept ='';
      for(let respStatus in this.props.path.responses){
        for(let acceptContentType in (this.props.path.responses[respStatus]["content"])){
          accept = accept + acceptContentType + ', '
        }
      }
      accept = accept.replace(/,\s*$/, ""); // remove trailing comma
      this.setState({expanded: true, accept: accept});
    }
  }

  render() {
    return (
    <div  className={'m-endpoint regular-font ' + this.props.path.method + (this.state.expanded?' expanded':' collapsed')}>
      <div onClick={ this.toggleExpand } className={'head ' + this.props.path.method +  (this.state.expanded?' expanded':' collapsed')}>
        <div className={'method ' + this.props.path.method} > {this.props.specType === 'rest' ? this.props.path.method : 'Method' } </div>
        <div className={'path ' + (this.props.path.deprecated?' deprecated':'')}> {this.props.path.path} </div>
        {this.props.path.deprecated && <span style={{ fontSize: '12px', textTransform: 'uppercase', fontWeight: 'bold', color: 'orangered', margin: '2px 0 0 5px' }}> deprecated </span>}
        <div className="only-large-screen" style={{ minWidth: '60px', flex: '1' }}></div>
        <div className="m-markdown-small descr"> {this.props.path.summary?<DangerousHTML html={marked(this.props.path.summary)} /> : ''}</div>
      </div>

      {this.state.expanded && (
      <div className={'body ' + this.props.path.method}>
        {(this.props.path.summary || this.props.path.description) && (
          <div className="summary">
            <div className="m-markdown title"><DangerousHTML html={marked(this.props.path.summary)} /></div>
            {this.props.path.summary !== this.props.path.description && (
              <div className="m-markdown">
                {this.props.path.description?<DangerousHTML html={marked(this.props.path.description)} />:''}
              </div>
            )}
          </div>
        )}
        <div className='req-resp-container'>
          <ApiRequest
            specType = {this.props.specType}
            method = {this.props.specType === 'rest' ? this.props.path.method : 'post'}
            path = {this.props.path.path}
            selectedServer = {this.props.selectedServer}
            parameters = {this.props.path.parameters}
            requestBody = {this.props.path.requestBody}
            accept ={this.state.accept}
          ></ApiRequest>
          <ApiResponse
            responses={this.props.path.responses}
            specType={this.props.specType}
          ></ApiResponse>
        </div>
      </div>
      )}

    </div>
    )
  }

}
