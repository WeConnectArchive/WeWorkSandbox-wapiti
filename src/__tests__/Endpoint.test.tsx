import * as React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { EndPoint } from '../EndPoint';

const thePath = {"path":"/test","description":"description","method":"GET",summary:"summary",responses:{"200":{"test":"test"}},parameters:{"test":"test"}}
const theType = "rest"

describe('Endpoint', () => {
  it('renders with rest', () => {
    let wrapper = shallow(<EndPoint
              selectedServer  = ""
              path           = {thePath}
              specType        = "rest"
            ></EndPoint>);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
  it('renders with rest expanded', () => {
    let wrapper = shallow(<EndPoint
              selectedServer  = ""
              path           = {thePath}
              specType        = "rest"
            ></EndPoint>);
    wrapper.setState({ expanded: true });
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });


});
