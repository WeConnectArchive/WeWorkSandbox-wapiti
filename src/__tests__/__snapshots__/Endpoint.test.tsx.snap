// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Endpoint renders with rest 1`] = `
<div
  className="m-endpoint regular-font GET collapsed"
>
  <div
    className="head GET collapsed"
    onClick={[Function]}
  >
    <div
      className="method GET"
    >
       
      GET
       
    </div>
    <div
      className="path "
    >
       
      /test
       
    </div>
    <div
      className="only-large-screen"
      style={
        Object {
          "flex": "1",
          "minWidth": "60px",
        }
      }
    />
    <div
      className="m-markdown-small descr"
    >
       
      <DangerousHTML
        className=""
        html="<p>summary</p>
"
        tagName="div"
      />
    </div>
  </div>
</div>
`;

exports[`Endpoint renders with rest expanded 1`] = `
<div
  className="m-endpoint regular-font GET expanded"
>
  <div
    className="head GET expanded"
    onClick={[Function]}
  >
    <div
      className="method GET"
    >
       
      GET
       
    </div>
    <div
      className="path "
    >
       
      /test
       
    </div>
    <div
      className="only-large-screen"
      style={
        Object {
          "flex": "1",
          "minWidth": "60px",
        }
      }
    />
    <div
      className="m-markdown-small descr"
    >
       
      <DangerousHTML
        className=""
        html="<p>summary</p>
"
        tagName="div"
      />
    </div>
  </div>
  <div
    className="body GET"
  >
    <div
      className="summary"
    >
      <div
        className="m-markdown title"
      >
        <DangerousHTML
          className=""
          html="<p>summary</p>
"
          tagName="div"
        />
      </div>
      <div
        className="m-markdown"
      >
        <DangerousHTML
          className=""
          html="<p>description</p>
"
          tagName="div"
        />
      </div>
    </div>
    <div
      className="req-resp-container"
    >
      <ApiRequest
        accept=""
        method="GET"
        parameters={
          Object {
            "test": "test",
          }
        }
        path="/test"
        selectedServer=""
        specType="rest"
      />
      <ApiResponse
        responses={
          Object {
            "200": Object {
              "test": "test",
            },
          }
        }
        specType="rest"
      />
    </div>
  </div>
</div>
`;
